lib/puppet/util/log.rb

```diff
@@ -170,6 +170,17 @@ def Log.with_destination(destination, &block)
   def Log.newmessage(msg)
     return if @levels.index(msg.level) < @loglevel
 
+    if msg.message.valid_encoding?
+      message = Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.message)
+    else
+      message = _("Received a log message with invalid encoding:")
+      message << Puppet::Util::CharacterEncoding.convert_to_utf_8(msg.message.dump) << "\n"
+      # We only select the last 5 callers in the stack to avoid being spammy
```

Comment:
I was actually looking at the bounds problem initially... it's not actually a problem, but you could also use `.first` for the same thing:

```
[1] pry(main)> foo = ['hi']
=> ["hi"]
[2] pry(main)> foo[0..4]
=> ["hi"]
[3] pry(main)> foo[0..4].join("\n")
=> "hi"
```